dbms/src/Interpreters/Join.cpp

```diff
@@ -916,13 +935,21 @@ Block Join::doJoinBlockHash(ProbeProcessInfo & probe_process_info) const
 
     bool enable_spill_join = isEnableSpill();
     JoinBuildInfo join_build_info{enable_fine_grained_shuffle, fine_grained_shuffle_count, enable_spill_join, is_spilled, build_concurrency, restore_round};
-    JoinPartition::probeBlock(partitions, rows, key_columns, key_sizes, added_columns, null_map, filter, current_offset, offsets_to_replicate, right_indexes, collators, join_build_info, probe_process_info);
+    JoinPartition::probeBlock(partitions, rows, key_columns, key_sizes, added_columns, null_map, filter, current_offset, offsets_to_replicate, right_indexes, collators, join_build_info, probe_process_info, flag_mapped_entry_helper_column);
     FAIL_POINT_TRIGGER_EXCEPTION(FailPoints::random_join_prob_failpoint);
+    /// For RIGHT_SEMI/RIGHT_ANTI join without other conditions, hash table has been marked already, just return empty build table header
+    if (isRightSemiFamily(kind) && !flag_mapped_entry_helper_column)
+    {
+        return sample_block_with_columns_to_add;
+    }
+
     for (size_t i = 0; i < num_columns_to_add; ++i)
     {
         const ColumnWithTypeAndName & sample_col = sample_block_with_columns_to_add.getByPosition(i);
         block.insert(ColumnWithTypeAndName(std::move(added_columns[i]), sample_col.type, sample_col.name));
     }
+    if (flag_mapped_entry_helper_column)
+        block.insert(ColumnWithTypeAndName(std::move(flag_mapped_entry_helper_column), flag_mapped_entry_helper_type, flag_mapped_entry_helper_name));
```

Comment:
No, it is only used locally.

Reply:
ok

Reply:
Currently, parameter of Join::Join() only contains left/right join keys, not all columns, so placed outside. And recheck its logic, find we should ensure its unique for both left and right sides, because we material the joined blocks after hash table lookup, and both sides' columns exist in the block.

Reply:
ok